### PR TITLE
WIP Update README.

### PR DESCRIPTION
**Pull request by nc6**
- Change to ReST format to include footnotes.

I want to add some more detail on extending the flow with user-defined effects.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+  steps. The `funflow-jobs <./funflow-jobs>`_ package provides support for
+  distributing full workflows using Redis. This has some extra constraints
+  compared to distributing external tasks, since Haskell functions are in
+  general not serialisable.
+- Safe and powerful workflow composition. FunFlow takes advantage of Haskell's
+  type system to ensure flows are composed in a safe manner. And since the
+  composition of flows results in a flow, it's easy to include one complex
+  workflow as part of another.
+
+Defining a workflow
```
Comment (bethaitman) [inline README.rst]: Test comment.

File: README.rst [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+  -- | A flow which doubles and reverses a list
+  doubleAndReverse :: SimpleFlow [Int] [Int]
+  doubleAndReverse = mapA (step (*2)) >>> step reverse
+
+The `step` function takes a Haskell function of type `a -> b` and lifts it into
+a `SimpleFlow a b`. [1]_ The `mapA` combinator applies a flow over a list, and
+`>>>` composes two flows sequentially.
+
+We can also use Haskell's arrow notation to compose more complex flows::
```
Comment (bethaitman) [inline README.rst]: We -> You

File: README.rst [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+  type system to ensure flows are composed in a safe manner. And since the
+  composition of flows results in a flow, it's easy to include one complex
+  workflow as part of another.
+
+Defining a workflow
+-------------------
+
+A simple workflow can be defined from basic Haskell functions::
+
+  -- | A flow which doubles and reverses a list
```
Comment (bethaitman) [inline README.rst]: "which doubles each element in a list, then reverses the list"

File: README.rst [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+  -- | Flow which asks for the user's name, then their favourite number.
+  --   The favourite number will be cached, such that if the user comes back,
+  --   they will not be asked again.
+  favouriteNumber = proc () -> do
+    name <- stepIO (\_ -> putStrLn "Hi, what's your name?" >> const getLine) -< ()
+    fnum <- stepIO' (def { cache = defaultCacherWithIdent 172652 })
+                    (\_ -> putStrLn "What's your favourite number?" >> const getLine) -< name
+    returnA -< (name, fnum)
+
+In this flow, we can also see the use of FunFlow's caching. `stepIO` lets us lift
```
Comment (bethaitman) [inline README.rst]: "In this flow, you can see FunFlow caching in use. `stepIO` lifts an `IO` action into a flow."

File: README.rst [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+
+  - An `in-memory <./funflow/src/Control/FunFlow/External/Coordinator/Memory.hs>`_
+    coordinator, useful for single-process computation.
+  - A `Redis based <./funflow/src/Control/FunFlow/External/Coordinator/Redis.hs>`_
+    coordinator.
+  - A `SQLite <./funflow/src/Control/FunFlow/External/Coordinator/SQLite.hs>`_
+    coordinator, which uses a SQLite database as a shared location for task
+    distribution.
+
+- Support for distributing complete workflows, rather than just individual
```
Comment (bethaitman) [inline README.rst]: Bold for the first clause here.

File: README.rst [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+  steps. The `funflow-jobs <./funflow-jobs>`_ package provides support for
+  distributing full workflows using Redis. This has some extra constraints
+  compared to distributing external tasks, since Haskell functions are in
+  general not serialisable.
+- Safe and powerful workflow composition. FunFlow takes advantage of Haskell's
```
Comment (bethaitman) [inline README.rst]: Bold for the first sentence here.

File: README.rst [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+  --   The favourite number will be cached, such that if the user comes back,
+  --   they will not be asked again.
+  favouriteNumber = proc () -> do
+    name <- stepIO (\_ -> putStrLn "Hi, what's your name?" >> const getLine) -< ()
+    fnum <- stepIO' (def { cache = defaultCacherWithIdent 172652 })
+                    (\_ -> putStrLn "What's your favourite number?" >> const getLine) -< name
+    returnA -< (name, fnum)
+
+In this flow, we can also see the use of FunFlow's caching. `stepIO` lets us lift
+an `IO` action into a flow. `stepIO'` is a more complex variant that lets us configure
```
Comment (bethaitman) [inline README.rst]: I'm never keen on "lets you do something" - it has the kind of odd sense of it giving you permission. Normally would replace with just the straightforward what-you-use-it-for - eg "(You can) use the more complex variant `stepIO` to configure things like the caching properties".

File: README.rst [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+wired together into a larger computation. Outputs from previous steps can form
+inputs to subsequent steps. 
+
+Features
+--------
+
+- Funflow is designed to be **integrated** into your application. It allows for
+  seamless composition of multiple types of computation:
+
+  - Pure functions (of type `a -> b`)
```
Comment (bethaitman) [inline README.rst]: Suggest bolding these too, for easy scanning.

File: README.rst [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+------------------
+
+To run a workflow, use one of the options from Control.FunFlow.Exec.Simple_ .
+The easiest place to start is with `withSimpleLocalRunner`. This takes the
+absolute path to a directory used to host the content store. It returns a
+function which can be used to run your workflow.
+
+Here's an example of a complete pipeline, along with a runner::
+
+  {-# LANGUAGE Arrows #-}
```
Comment (bethaitman) [inline README.rst]: Worth adding comments here?

File: README.rst [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+    res <- withSimpleLocalRunner [absdir|/tmp/funflow|] $ \run ->
+      run favouriteNumber ()
+    case res of
+      Left err -> putStrLn $ "Something went wrong: " ++ show err
+      Right (name, num) -> putStrLn $ "Hi, " ++ name ++ ", your favourite number is " ++ num
+
+Defining external tasks
+-----------------------
+
+External tasks are used to run steps outside of the Haskell process.
```
Comment (bethaitman) [inline README.rst]: "Use external tasks to run"

File: README.rst [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+Fundamentally, an external task will resolve to a command-line call to another
+application. This can obviously be done with `stepIO`, but using external tasks
```
Comment (bethaitman) [inline README.rst]: Suggest replacing "obviously", as it's a value judgement.

"While you could do this with `stepIO`, using external tasks..."

File: README.rst [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+has a couple of big advantages:
+
+- External tasks fit in naturally with the content store framework. Both the
+  inputs to the task and the actual definition of the task will be used to
```
Comment (bethaitman) [inline README.rst]: "will be" -> "are"

File: README.rst [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+  determine the resultant hash, so that if, say, a script changes, the
+  results will be recomputed. This is hard to achieve with `stepIO`.
+- External tasks are naturally distributable. When running in a production setting,
+  you're likely to want to distribute tasks among multiple machines. This is
+  not, in general, achievable with arbitrary `IO` computations.
```
Comment (bethaitman) [inline README.rst]: remove "arbitrary", it's tautologous

File: README.rst [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+
+FunFlow's current approach to external tasks is heavily based on Docker_. Using
+Docker allows tasks to be self-contained, and adds minimal requirements to the
+system being used to host funflow instances (they just need to have docker running!)
```
Comment (bethaitman) [inline README.rst]: you don't need that exclamation mark. And add a full stop after the bracket

File: README.rst [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+
+To use a docker container as an external step, one simply needs to define a function
```
Comment (bethaitman) [inline README.rst]: delete "one simply needs to"

File: README.rst [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+
+- Specifying the docker image (and optionally, image ID) to use. For example,
+  `nixos/nix:1.11.14`.
+- Specifying the path to the command which will be run within the container.
+- Specifying which inputs (from the content store) are to be mounted within
+  the container.
+
+Here's an example::
+
+  myDockerStep :: SimpleFlow CS.Item CS.Item
```
Comment (bethaitman) [inline README.rst]: This could do with comments too.

File: README.rst [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+    , Docker.args =
+        [
+        , "--input_dir", "/input/"
+        , "--output_dir", "/output/"
+        ]
+    }
+
+A `CS.Item` refers to an item within the content store. You can use
+`putInStore`, `getFromStore`, `copyFileToStore` and similar tools to add and
+fetch files from the store. This lets you interleave internal and external
```
Comment (bethaitman) [inline README.rst]: Feels like you could do with a code example of this too, to clarify.

File: README.rst [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+computations.
+
+Running on multiple machines
+----------------------------
+
+To run on multiple machines, you need to use one of the distributable
+coordinators - either the Redis coordinator or the SQLite coordinator. To do
+this, you need to:
+
+- Start some *executors* pointed at the coordinator.
```
Comment (bethaitman) [inline README.rst]: These should probably be numbered steps, as the order matters.

Comment (bethaitman) [inline README.rst]: What is an executor? Needs defining.

File: README.rst [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+- Run your flow using that coordinator.
+
+The simplest way to run an executor is to use the bundled `ffexecutord` executable.
+This can work with either the Redis or SQLite coordinators. Here's how we might
```
Comment (bethaitman) [inline README.rst]: I'd break the line at "Here's", because it introduces a new idea, so should be a new paragraph.

Comment (bethaitman) [inline README.rst]: "Here's an example of initialising"

File: README.rst [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+initialise an executor using `/tmp/funflow` as the content store directory, and
+`/tmp/coordinator.db` as our coordinating database::
+
+  ffexecutord sqlite /tmp/funflow /tmp/coordinator.db
+
+We then need to run the flow, pointing at this coordinator. To do so, we'll need a
```
Comment (bethaitman) [inline README.rst]: We -> You, throughout.

File: README.rst [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+slightly more complex function from Control.FunFlow.Exec.Simple_: `runSimpleFlow`. We
+need to give this the correct parameters for the `SQLite` coordinator::
```
Comment (bethaitman) [inline README.rst]: Not sure why "correct" is relevant here? Just "the parameters"?

File: README.rst [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+    runFlow SQLite [absfile|/tmp/coordinator.db|] store runNoEffect 123123 flow input
+
+A couple of the parameters here may be confusing:
+
+runNoEffect
+  This is used to handle any user-defined effects in the flow. Since
+  we have none here, we can use `runNoEffect`.
+
+123123
+  This is a random integer used in helping to determine the hashes for caching
```
Comment (bethaitman) [inline README.rst]: Not clear whether this is a number you randomise every time, or this is a number you've picked for now, or you need to use 123123 specifically here. Worth clarifying.

File: README.rst [hunk 23]
```diff
@@ -0,0 +1,227 @@
...
+We then need to run the flow, pointing at this coordinator. To do so, we'll need a
+slightly more complex function from Control.FunFlow.Exec.Simple_: `runSimpleFlow`. We
+need to give this the correct parameters for the `SQLite` coordinator::
+
+  CS.withStore [absdir|/tmp/funflow|] $ \store -> do
+    runFlow SQLite [absfile|/tmp/coordinator.db|] store runNoEffect 123123 flow input
+
+A couple of the parameters here may be confusing:
+
+runNoEffect
```
Comment (bethaitman) [inline README.rst]: I like the formatting you've used here for explaining the code snippet - could do the same with the first code snippet too, as you have a similar explanation after that.

File: README.rst [hunk 24]
```diff
@@ -0,0 +1,227 @@
+FunFlow
+=======
+
+.. highlight:: haskell
+.. default-role:: code
+
+.. image:: https://circleci.com/gh/tweag/funflow.svg?style=svg
+    :target: https://circleci.com/gh/tweag/funflow
+
+FunFlow is a library and tools to compose and run computational workflows.
```
Comment (bethaitman) [inline README.rst]: This is not super clear to me. I think it'd be helpful to explain a little about why you'd want to use it, or what benefits it'll bring - WIIFM, basically.

File: README.rst [hunk 25]
```diff
@@ -0,0 +1,227 @@
...
+
+FunFlow is a library and tools to compose and run computational workflows. 
+A workflow is a computation built up of multiple steps. These steps are then
+wired together into a larger computation. Outputs from previous steps can form
+inputs to subsequent steps. 
+
+Features
+--------
+
+- Funflow is designed to be **integrated** into your application. It allows for
```
Comment (bethaitman) [inline README.rst]: This feels like it's two separate points.

File: README.rst [hunk 26]
```diff
@@ -0,0 +1,227 @@
...
+
+.. image:: https://circleci.com/gh/tweag/funflow.svg?style=svg
+    :target: https://circleci.com/gh/tweag/funflow
+
+FunFlow is a library and tools to compose and run computational workflows. 
+A workflow is a computation built up of multiple steps. These steps are then
+wired together into a larger computation. Outputs from previous steps can form
+inputs to subsequent steps. 
+
+Features
```
Comment (bethaitman) [inline README.rst]: Worth reviewing the ordering of these - making sure you're sure the most important ones come first.

File: README.rst [hunk 27]
```diff
@@ -0,0 +1,227 @@
...
+--------
+
+- Funflow is designed to be **integrated** into your application. It allows for
+  seamless composition of multiple types of computation:
+
+  - Pure functions (of type `a -> b`)
+  - IO actions (of type `a -> IO b`)
+  - *External* computations. External computations are executed outside of the
```
Comment (bethaitman) [inline README.rst]: "the Haskell process" which process? You haven't mentioned this before. "Outside FunFlow"?

File: README.rst [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+    Haskell process. They allow funflow to invoke computations from any
```
Comment (bethaitman) [inline README.rst]: Consistent capitalisation of FunFlow pls :)

File: README.rst [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+    language.
+  - User-defined actions. FunFlow is extensible with user-defined actions. By
```
Comment (bethaitman) [inline README.rst]: Not quite sure what an action is.

Comment (bethaitman) [inline README.rst]: I'd switch these clauses around - start with the benefit ie mocking

File: README.rst [hunk 30]
```diff
@@ -0,0 +1,227 @@
...
+    specifying different interpreters for such actions, you can easily test
+    funflow in a mock environment.
+
+- Funflow provides very powerful **caching** for steps.
+
+  - FunFlow's caching is based around the *content store*. Inspired by the nix_
```
Comment (bethaitman) [inline README.rst]: The `nix` sentence feels like an aside. I'd put it after the benefit, ie lead with "This means Funflow knows when to re-run a computation and when to re-use a previous result."

File: README.rst [hunk 31]
```diff
@@ -0,0 +1,227 @@
...
+    package manager, this links all artifacts according to the hash of the
+    inputs and computations which produced them. FunFlow uses this to know when
+    exactly a computation must be rerun and when previous results can be re-used.
+  - The content store also acts as a CAS_ system. This means that, if multiple
+    inputs produce the same output, that output will be stored only once on disk
+    (and subsequent computations will not be rerun). This is particularly useful
+    for build systems: if you change your input in such a way that the ouptut
```
Comment (bethaitman) [inline README.rst]: We discussed this sentence could do with rewriting to make it clearer.

File: README.rst [hunk 32]
```diff
@@ -0,0 +1,227 @@
...
+    is not altered, there is no need to run the rest of the computation again.
+  - Content store caching is used by default for all external steps, and can be
+    enabled for internal computations by providing suitable
```
Comment (bethaitman) [inline README.rst]: "and can be enabled for internal computations if you provide..."

File: README.rst [hunk 33]
```diff
@@ -0,0 +1,227 @@
...
+    serialisation/deserialisation functions.
+
+- **Failure handling**. In a long running workflow, failures are inevitable. FunFlow
```
Comment (bethaitman) [inline README.rst]: long-running

File: README.rst (hunk 25 above)
Comment (nc6) [reply README.rst]: Fixed

**Pull request merged**